### PR TITLE
[TRIE] add trie index type in DataSourceMeta

### PR DESCRIPTION
**Pull request by adrian-wang**
## What changes were proposed in this pull request?

add trie index type in DataSourceMeta. add a config value of "spark.sql.oap.oindex.trie.enabled" to create permuterm index

## How was this patch tested?

unit tests.

### CODE REVIEW

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 1]
```diff
@@ -105,7 +105,17 @@ case class CreateIndex(
         metaBuilder.withNewSchema(schema)
       }
 
-      indexType match {
+      val innerIndexType =
+        if (sparkSession.conf.get(SQLConf.OAP_ENABLE_TRIE_OVER_BTREE) &&
+          indexType == BTreeIndexType && indexColumns.length == 1) {
+          val entry = schema.map(_.name).toIndexedSeq.indexOf(indexColumns(0).columnName)
+          if (schema(entry).dataType == StringType) {
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: Add a comment in the code

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 2]
```diff
@@ -105,7 +105,17 @@ case class CreateIndex(
         metaBuilder.withNewSchema(schema)
       }
 
-      indexType match {
+      val innerIndexType =
+        if (sparkSession.conf.get(SQLConf.OAP_ENABLE_TRIE_OVER_BTREE) &&
+          indexType == BTreeIndexType && indexColumns.length == 1) {
+          val entry = schema.map(_.name).toIndexedSeq.indexOf(indexColumns(0).columnName)
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: `schema(indexColumns(0).columnName).dataType.sameType(StringType)` ?

File: src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala [hunk 3]
```diff
@@ -116,6 +126,10 @@ case class CreateIndex(
           val entries = indexColumns.map(col =>
             schema.map(_.name).toIndexedSeq.indexOf(col.columnName))
           metaBuilder.addIndexMeta(new IndexMeta(indexName, time, BitMapIndex(entries)))
+        case PermutermIndexType =>
+          val entries = indexColumns.map(col =>
```
Comment (chenghao-intel) [inline src/main/scala/org/apache/spark/sql/execution/datasources/oap/index/indexPlans.scala]: `indexColumns(0)...`? Since we only have a single key here.

**Comment by adrian-wang** [comment]
thanks @chenghao-intel , I will refine this pr tonight accordingly, and merge this before Saturday ends, so that other prs will not be blocked.

**Pull request merged**